python/lsst/daf/butler/formatters/pickle.py

```diff
@@ -34,10 +34,35 @@
 import pickle
 from typing import Any
 
+from lsst.resources import ResourcePath
+
 from .file import FileFormatter
+from .typeless import TypelessFormatter
+
+
+class PickleFormatter(TypelessFormatter):
+    """Interface for reading and writing Python objects to and from pickle
+    files.
+    """
+
+    default_extension = ".pickle"
+    unsupported_parameters = None
+    can_read_from_uri = True
+
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None) -> Any:
+        # Read the pickle file directly from the resource into memory.
+        try:
+            data = pickle.loads(uri.read())
+        except pickle.PicklingError:
+            data = None
+
+        return data
```

Comment:
For a simple formatter like this one, where all we need is bytes, I was expecting it'd be better practice to implement `read_from_stream` rather than `read_from_uri`, because I think of the former as more general.  I can see why others might think the opposite, though, since we could in theory add a `ResourcePath` implementation for anything, even though there are stream types for which there is at present no `ResourcePath` implementation (e.g. `io.BytesIO`).  It's not a big deal because I think either is fine, but it might be good to think about what we recommend when there doesn't seem to be an obvious reason to prefer one or the other.

Reply:
I haven't tried to see if that works specifically for pickles. It definitely doesn't work for many other cases because `readline` is not implemented by `ResourcePath`.

Reply:
I checked and `pickle.load()` requires a `readline` implementation. I'm starting to think we are going to have to implement that in ResourcePath.

Reply:
`yaml.load` does seem to work with ResourcePath though it 's much worse performance because when you hand a stream to `yaml.load` it does reads in chunks of 4096 bytes. There is a 8MB yaml file in summit_utils. If I read that up front and then parse it it completes in 3 seconds but if I pass the handle to the loader it takes 24 seconds so reading up front is definitely better.